engine/execution/provider/engine_test.go

```diff
@@ -88,16 +119,22 @@ func TestProviderEngine_onChunkDataRequest(t *testing.T) {
 
 		ps.On("AtBlockID", blockID).Return(ss)
 		ss.On("Identity", originID).Return(unittest.IdentityFixture(unittest.WithRole(flow.RoleExecution), unittest.WithWeight(0)), nil)
-		execState.On("ChunkDataPackByChunkID", mock.Anything, mock.Anything).Return(chunkDataPack, nil)
+		execState.On("ChunkDataPackByChunkID", mock.Anything).Return(chunkDataPack, nil)
 		execState.On("GetBlockIDByChunkID", chunkID).Return(blockID, nil)
 
 		req := &messages.ChunkDataRequest{
 			ChunkID: chunkID,
 			Nonce:   rand.Uint64(),
 		}
+		cancelCtx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		ctx, _ := irrecoverable.WithSignaler(cancelCtx)
+		e.Start(ctx)
 		// submit using origin ID with zero weight
 		unittest.RequireCloseBefore(t, e.Ready(), 100*time.Millisecond, "could not start engine")
-		e.onChunkDataRequest(context.Background(), originID, req)
+		require.NoError(t, e.Process(channels.RequestChunks, originID, req))
+		time.Sleep(1 * time.Second) // waits enough so that request is picked from queue.
```

Comment:
It would be nice to avoid needing to sleep in all these tests. I don't think we have a general approach for waiting until a particular message has been processed, but here we could drain in-progress requests before shutting down the engine. Suggestion below (using terminology from TODO link):
* The `RequestShoveler` thread listens for `context.Done()` and (1) closes the `RequestChannel` (2) exits
* The `RequestProcessor` threads range over the `RequestChannel` channel (do not listen for `context.Done()` and exit the range statement when the `RequestChannel` is emptied and closed

Then we will drain and process all the requests retrieved from the `MessageStore` before terminating the engine. This would be useful for testing, because we could:
* Call `Process` on some message
* Assert that `engine.MessageStore` is empty
* Call `cancel` to drain then terminate the engine - we are sure that the request will be processed before the engine shuts down, and we don't need to wait.

Reply:
https://github.com/onflow/flow-go/pull/2951/commits/4b92f13885fa9eeb12a894b6d0f9773ab74f1959